src/Controller/ControllerFactory.php

```diff
@@ -156,58 +156,83 @@ public function handle(ServerRequestInterface $request): ResponseInterface
-            // If there is no type we can't look in the container
-            // assume the parameter is a passed param
-            $type = $parameter->getType();
-            if (!$type) {
-                if (count($passedParams)) {
-                    $args[$position] = array_shift($passedParams);
-                } elseif ($hasDefault) {
-                    $args[$position] = $parameter->getDefaultValue();
+                if (is_string($argument)) {
+                    $resolved[] = $this->coerceStringToType($argument, $typeName);
+                } else {
+                    $resolved[] = $argument;
                 }
                 continue;
             }
 
-            $typeName = $type instanceof ReflectionNamedType ? ltrim($type->getName(), '?') : null;
+            // Check for dependency injection
+            if ($typeName) {
+                if ($this->container->has($typeName)) {
+                    $resolved[] = $this->container->get($typeName);
+                    continue;
+                }
+            }
 
-            // Primitive types are passed args as they can't be looked up in the container.
-            // We only handle strings currently.
-            if ($typeName === 'string') {
-                if (count($passedParams)) {
-                    $args[$position] = array_shift($passedParams);
-                } elseif ($hasDefault) {
-                    $args[$position] = $parameter->getDefaultValue();
+            // Default to positional parameter
+            if ($passedParams) {
+                $argument = array_shift($passedParams);
+                if (is_string($argument)) {
+                    $resolved[] = $this->coerceStringToType($argument, $typeName);
+                } else {
+                    $resolved[] = $argument;
                 }
                 continue;
             }
 
-            // Check the container and parameter default value.
-            if ($typeName && $this->container->has($typeName)) {
-                $args[$position] = $this->container->get($typeName);
-            } elseif ($hasDefault) {
-                $args[$position] = $parameter->getDefaultValue();
+            // Add default value if none provided
+            if ($parameter->isDefaultValueAvailable()) {
+                $resolved[] = $parameter->getDefaultValue();
+                continue;
             }
 
-            if (!array_key_exists($position, $args)) {
-                throw new InvalidArgumentException(
-                    "Could not resolve action argument `{$parameter->getName()}`. " .
-                    'It has no definition in the container, no passed parameter, and no default value.'
-                );
+            // Variadic parameter can have 0 arguments
+            if ($parameter->isVariadic()) {
+                continue;
             }
+
+            throw new MissingActionException(
+                "Unable to find argument for action `{$function->getName()}` parameter `$name`."
```

Comment:
```suggestion
                "Unable to find argument for action `{$function->getName()}` parameter `{$name}`."
```

Little bit clearer that this is variable interpolation.